docs/manual/src/internals/design_principles.md

```diff
@@ -0,0 +1,95 @@
+
+
+### Performance is a feature, but not a deal-breaker
+
+Our initial use-cases are not performance-critical, and our team are not low-level Rust experts,
+so we're highly motivated to favour simplicity and maintainability over performance. Given the choice
+we will pick "simple but slow" over "fast but complicated".
+
+However, we know that performance can degrade through thousands of tiny cuts, so we'll keep iterating
+towards the winning combination of "simple *and* fast" over time.
+
+**Example:** Initial versions of the tool used opaque integer handles and explicit mutexes to manage
+object references, favouring simplicity (in the "we're confident this works as intended" sense) over
+performance. As we got more experience and confidence with the approach tool we replaced handles with
+raw `Arc` pointers, which both simplified the code and removed some runtime overheads.
+
+**Violation:** Initial versions of the tool pass structured data over the FFI by serializing it to
+a byte buffer, favouring ease of implementation and understanding over performance. However, we have
+not done any work to measure the performace impact or iterate towards something with lower overhead
+(such as using `repr(C)` structs).
+
+
+### Produce bindings that feel idiomatic for the target language
+
+The generated bindings should feel idiomatic for their end users, and what feels idiomatic can differ
+between different target languages. Ideally consumers should not even realize that they're using
+bindings to Rust under the hood.
+
+We'll accept extra complexity inside of UniFFI if it means producing bindings that are nicer for consumers to use.
+
+**Example:** We case-convert names to match the accepted standards of the target language,
+so a method named `do_the_thing` in Rust might be called `doTheThing` in its Kotlin bindings.
+
+**Example:** Object references try to integrate with the GC of the target language, so that holding
+a reference to a Rust struct feels like holding an ordinary object instance.
+
+**Violation:** The Kotlin bindings have an explicit `destroy` method on object instances, because we haven't
+yet found a good way to integrate with the JVM's GC.
+
+
+### Empower users to debug and maintain the tool
+
+To succeed long-term, we can't depend on a dedicated team of "UniFFI experts" for debugging and maintenance.
+The people using the tool need to be empowered to debug, maintain and develop it.
+
+If you're using UniFFI-generated bindings and something doesn't work quite right, it should be possible
+for you to dig in to the generated foreign-language code, follow it through to the underlying Rust code,
+and work out what's going wrong without being an expert in Rust or UniFFI.
+
+**Example:** We try to include comments in the generated code to help guide users who may be reading
+through it to debug some issue.
+
+**Example:** We use a separate interface description file rather than annotating the Rust code directly
```

Comment:
I don't really get this point, not how the UDL helps "trace the flow of data" and I think the UDL actually works against that - understanding weedle to maintain the tool is a liability rather than a feature. The UDL is also a liability in other ways, such as limiting ways we can express concepts uniffi should have.  This point would make sense if there was a section titled something like "move fast, make pragmatic decisions" or similar, but there's not (and I don't really think it needs one).

Reply:
Thanks, I will just delete this one.